packages/block-editor/src/components/block-tools/index.js

```diff
@@ -1,14 +1,46 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useViewportMatch } from '@wordpress/compose';
+import { Popover } from '@wordpress/components';
+
 /**
  * Internal dependencies
  */
 import InsertionPoint from './insertion-point';
 import BlockPopover from './block-popover';
+import { store as blockEditorStore } from '../../store';
+import BlockContextualToolbar from './block-contextual-toolbar';
 
+/**
+ * Renders block tools (the block toolbar, select/navigation mode toolbar, the
+ * insertion point and a slot for the inline rich text toolbar). Must be wrapped
+ * around the block content and editor styles wrapper or iframe.
+ *
+ * @param {Object} $0          Props.
+ * @param {Object} $0.children The block content and style container.
+ */
 export default function BlockTools( { children } ) {
+	const isLargeViewport = useViewportMatch( 'medium' );
+	const hasFixedToolbar = useSelect(
+		( select ) => select( blockEditorStore ).getSettings().hasFixedToolbar,
+		[]
+	);
+
 	return (
 		<InsertionPoint>
+			{ ( hasFixedToolbar || ! isLargeViewport ) && (
```

Comment:
In an earlier iteration this component was just implementing the two forms of the block toolbar, and I think it was good to include the viewport logic in that iteration. 

But now this is bundled with some pretty key functionality for the block editor it seems fairly opinionated to say that someone implementing an editor should always switch to a fixed toolbar on small viewports.

I don't think it's a blocker, this PR still makes things much easier overall. And an implementer can still reimplement this component without the viewport code. But I think it does pose a question about whether `BlockEditor` should be exporting high-level or low-level components.

Reply:
We can have a prop on this component `hasFixedToolbar` that overrides the settings?

Reply:
No rush to do that right now, just some food for thought. 😄